File: tests/fixtures/python-simple/src/payments/service.py
"""Payment service — handles payment processing."""

from dataclasses import dataclass
from typing import Optional

from .types import PaymentResult, CardDetails
from ..utils.logger import Logger


class PaymentService:
    """Handles payment processing."""

    def __init__(self, client, logger: Logger):
        self._client = client
        self._logger = logger

    async def process_payment(self, amount: float, user_id: str) -> PaymentResult:
        """Process a payment for the given user."""
        self._logger.info(f"Processing payment for {user_id}")
        return await self._client.charge(amount, user_id)

    def refund(self, tx_id: str, reason: Optional[str] = None) -> bool:
        """Refund a previous transaction."""
        return self._client.refund(tx_id, reason)

    @staticmethod
    def validate_card(card: CardDetails) -> bool:
        """Validate a credit card."""
        return card.number is not None

    @property
    def is_connected(self) -> bool:
        """Check if payment client is connected."""
        return self._client.connected

    def _calculate_fee(self, amount: float) -> float:
        """Internal: calculate processing fee."""
        return amount * 0.029

    def __apply_discount(self, amount: float, code: str) -> float:
        """Name-mangled: apply a discount code."""
        return amount * 0.9

File: tests/fixtures/rust-simple/src/payments/service.rs
use crate::payments::types::{CardDetails, PaymentResult};
use crate::utils::logger::Logger;

/// Handles payment processing.
pub struct PaymentService {
    client: Box<dyn PaymentClient>,
    logger: Logger,
}

impl PaymentService {
    /// Create a new payment service.
    pub fn new(client: Box<dyn PaymentClient>, logger: Logger) -> Self {
        Self { client, logger }
    }

    /// Process a payment for a given user.
    pub async fn process_payment(&self, amount: f64, user_id: &str) -> PaymentResult {
        self.logger.info(&format!("Processing payment for {user_id}"));
        self.client.charge(amount, user_id).await
    }

    /// Refund a transaction.
    pub fn refund(&self, tx_id: &str, reason: Option<&str>) -> bool {
        self.logger.info("Processing refund");
        self.client.refund(tx_id, reason)
    }

    fn validate_card(card: &CardDetails) -> bool {
        card.number.is_some()
    }
}

/// Trait for payment processing backends.
pub trait PaymentClient: Send + Sync {
    /// Charge a user.
    async fn charge(&self, amount: f64, user_id: &str) -> PaymentResult;
    /// Refund a transaction.
    fn refund(&self, tx_id: &str, reason: Option<&str>) -> bool;
}

/// A mock payment client for testing.
pub struct MockPaymentClient;

impl PaymentClient for MockPaymentClient {
    async fn charge(&self, _amount: f64, _user_id: &str) -> PaymentResult {
        PaymentResult::Success {
            tx_id: "mock-tx-001".to_string(),
        }
    }

    fn refund(&self, _tx_id: &str, _reason: Option<&str>) -> bool {
        true
    }
}
